File: src/main/java/baekjoon/dp/합_분해.java
package baekjoon.dp;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

/**
 * https://www.acmicpc.net/problem/2225
 *
 * 정수 k 개의 합으로 n 을 구하는 방법은
 * 정수 k-1 개의 합으로 n-i 를 구하는 합과 같다
 * n-i 에다가 i를 더해주면 되기 떄문이다(k 1개 증가)
 * > fn(k)(n) = fn(k-1)(n-i) // 0 <= i <= n
 *
 * [4][6] = [3][0]+[3][1]+[3][2]+[3][3]+[3][4]+[3][5]+[3][6]
 * > 0 과 6을 더해주는건 0 + 6, 6 + 0 도 연산에 포함되기 때문이다
 */
public class 합_분해 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        합_분해 fn = new 합_분해();
        int[] param = Arrays.stream(br.readLine().split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();
        System.out.println(fn.operationCount(param[1], param[0]));
    }

    private int mod = 1_000_000_000;

    public int operationCount(int k, int n) {
        int[][] dp = new int[k + 1][n + 1];

        dp[0][0] = 1;
        for (int i = 1; i <= k; i++) {
            for (int j = 0; j <= n; j++) {
                for (int l = 0; l <= j; l++) {
                    dp[i][j] += dp[i - 1][l] % mod;
                }
            }
        }

        return dp[k][n] % mod;
    }
}

File: src/main/java/baekjoon/dp/basic/합분해.kt
package baekjoon.dp.basic

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader

/**
 * https://www.acmicpc.net/problem/2225
 *
 * fn(n, k) : k개의 숫자로 n을 만들기 위한 경우의 수
 * fn(n, k)를 구하려면,
 * 첫번째 피연산자로 0부터 n까지 오고, 이후 피연산자는 fn(n-i, k-1)으로 올 수 있는 숫자들의 묶음이 오면 된다
 * 경우의 수를 더하는 것이므로 위의 문장을 하나의 연산으로 보면 되니까..
 * 최종 점화식 : fn(n, k) = fn(n-i, k-1) // 0 <= i <= n
 *
 * fn(10, 3)을 구해야한다고 가정해보면,
 * fn(10, 3) = (0 + fn(10, 2)) + (1 + fn(9, 2)) + (2 + fn(8, 2)) + (3 + fn(7, 2)) ...
 * 이 되는 것이다
 *
 * 0 + fn(10, 2) 를 또 풀어서 보면
 * 0 + 1 + 9
 * 0 + 2 + 8
 * 0 + 3 + 7
 * ...
 * 의 형태가 된다
 *
 */
fun main(args: Array<String>) {
    println(합분해().solution(10, 2) == 11)
    println(합분해().solution(20, 2) == 21)
    println(합분해().solution(3, 3) == 10)
    println(합분해().solution(1, 1) == 1)
    println(합분해().solution(2, 1) == 1)

    try {
        val fn = 합분해()
        val br = BufferedReader(InputStreamReader(System.`in`))
        val args = br.readLine().split(" ")
        println(fn.solution(args[0].toInt(), args[1].toInt()))
    } catch (e: IOException) {/* */}
}

class 합분해 {
    val mod = 1_000_000_000

    fun solution(number: Int, count: Int):Int {
        val dp = Array(number + 1) { IntArray(count + 1) }

        for (i in 0..number) {
            dp[i][1] = 1
        }

        for (i in 2..count) {
            for (j in 0..number) {
                for (k in 0..j) {
                    dp[j][i] += dp[j - k][i - 1]
                    dp[j][i] %= mod
                }
            }
        }

        return dp[number][count] % mod
    }
}